modules/queue/queue_disk.go

```diff
@@ -0,0 +1,207 @@
+		closed:     make(chan struct{}),
+		terminated: make(chan struct{}),
+		workers:    config.Workers,
+		name:       config.Name,
+	}
+	queue.pool.qid = GetManager().Add(queue, LevelQueueType, config, exemplar, queue.pool)
+	return queue, nil
+}
+
+// Run starts to run the queue
+func (l *LevelQueue) Run(atShutdown, atTerminate func(context.Context, func())) {
+	atShutdown(context.Background(), l.Shutdown)
+	atTerminate(context.Background(), l.Terminate)
+
+	go func() {
+		_ = l.pool.AddWorkers(l.workers, 0)
+	}()
+
+	go l.readToChan()
+
+	log.Trace("LevelQueue: %s Waiting til closed", l.name)
+	<-l.closed
+
+	log.Trace("LevelQueue: %s Waiting til done", l.name)
+	l.pool.Wait()
+
+	log.Trace("LevelQueue: %s Waiting til cleaned", l.name)
+	ctx, cancel := context.WithCancel(context.Background())
+	atTerminate(ctx, cancel)
+	l.pool.CleanUp(ctx)
+	cancel()
+	log.Trace("LevelQueue: %s Cleaned", l.name)
+
+}
+
+func (l *LevelQueue) readToChan() {
+	for {
+		select {
+		case <-l.closed:
+			// tell the pool to shutdown.
+			l.pool.cancel()
+			return
+		default:
+			bs, err := l.queue.RPop()
+			if err != nil {
+				if err != levelqueue.ErrNotFound {
+					log.Error("LevelQueue: %s Error on RPop: %v", l.name, err)
+				}
+				time.Sleep(time.Millisecond * 100)
+				continue
+			}
+
+			if len(bs) == 0 {
+				time.Sleep(time.Millisecond * 100)
+				continue
+			}
+
+			var data Data
+			if l.exemplar != nil {
+				t := reflect.TypeOf(l.exemplar)
+				n := reflect.New(t)
+				ne := n.Elem()
+				err = json.Unmarshal(bs, ne.Addr().Interface())
+				data = ne.Interface().(Data)
+			} else {
+				err = json.Unmarshal(bs, &data)
+			}
+			if err != nil {
+				log.Error("LevelQueue: %s Failed to unmarshal with error: %v", l.name, err)
+				time.Sleep(time.Millisecond * 100)
+				continue
+			}
+
+			log.Trace("LevelQueue %s: Task found: %#v", l.name, data)
+			l.pool.Push(data)
+			time.Sleep(time.Millisecond * 10)
+
+		}
+	}
+}
+
+// Push will push the indexer data to queue
+func (l *LevelQueue) Push(data Data) error {
+	if l.exemplar != nil {
+		// Assert data is of same type as r.exemplar
+		value := reflect.ValueOf(data)
+		t := value.Type()
+		exemplarType := reflect.ValueOf(l.exemplar).Type()
+		if !t.AssignableTo(exemplarType) || data == nil {
+			return fmt.Errorf("Unable to assign data: %v to same type as exemplar: %v in %s", data, l.exemplar, l.name)
+		}
+	}
+	bs, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+	return l.queue.LPush(bs)
+}
+
+// Shutdown this queue and stop processing
+func (l *LevelQueue) Shutdown() {
+	log.Trace("LevelQueue: %s Shutdown", l.name)
+	select {
+	case <-l.closed:
+	default:
+		close(l.closed)
+	}
+}
+
+// Terminate this queue and close the queue
+func (l *LevelQueue) Terminate() {
+	log.Trace("LevelQueue: %s Terminating", l.name)
+	l.Shutdown()
+	select {
+	case <-l.terminated:
+	default:
+		close(l.terminated)
+		if err := l.queue.Close(); err != nil && err.Error() != "leveldb: closed" {
```

Comment:
Problem here is we will not know if the error description changes in a future library update.

Reply:
Don't really have a choice - it's either that or pass it up and log it unnecessarily
 
We'll just have to be aware of it.